File: backend/app/services/events.py
import re
from app.core.errors import invariant
from app.core.store import new_id, now, store
from app.schemas import Event, EventCreate, EventStatus, TicketCategory, TicketCategoryCreate
from app.services.audit import log_audit


def slugify(title: str) -> str:
    return re.sub(r"(^-|-$)", "", re.sub(r"[^a-z0-9]+", "-", title.lower()))


def list_events() -> list[Event]:
    return store.events


def get_event_by_slug(slug: str) -> Event | None:
    return next((event for event in store.events if event.slug == slug or event.id == slug), None)


def create_event(payload: EventCreate) -> Event:
    invariant(any(org.id == payload.organization_id for org in store.organizations), 404, "ORGANIZATION_NOT_FOUND", "Organization not found")
    event = Event(
        id=new_id("evt"),
        organization_id=payload.organization_id,
        title=payload.title,
        slug=slugify(payload.title),
        description=payload.description,
        event_type=payload.event_type,
        venue=payload.venue,
        city=payload.city,
        start_time=payload.start_time,
        end_time=payload.end_time,
        status=EventStatus.DRAFT,
        visibility=payload.visibility,
        capacity=payload.capacity,
        created_at=now(),
    )
    store.events.append(event)
    log_audit("event.created", "event", event.id, organization_id=event.organization_id, event_id=event.id)
    return event


def list_ticket_categories(event_id: str | None = None) -> list[TicketCategory]:
    return [cat for cat in store.ticket_categories if event_id is None or cat.event_id == event_id]


def create_ticket_category(payload: TicketCategoryCreate) -> TicketCategory:
    invariant(any(event.id == payload.event_id for event in store.events), 404, "EVENT_NOT_FOUND", "Event not found")
    category = TicketCategory(
        id=new_id("cat"),
        organization_id=payload.organization_id,
        event_id=payload.event_id,
        name=payload.name,
        price_paisa=payload.price_paisa,
        capacity=payload.capacity,
        created_at=now(),
    )
    store.ticket_categories.append(category)
    log_audit("ticket_category.created", "ticket_category", category.id, organization_id=category.organization_id, event_id=category.event_id)
    return category

File: backend/modules/events.ts
import { createId } from "../core/ids";
import { getStore, persistStoreRecord, persistStoreUpdate } from "../core/store";

export type EventRecord = {
  id: string;
  organizationId: string;
  slug: string;
  title: string;
  description: string;
  eventType: string;
  status: string;
  visibility: string;
  venue: string;
  city: string;
  startTime: string;
  gpsRequired: boolean;
  [key: string]: unknown;
};

export type TicketCategoryRecord = {
  id: string;
  eventId: string;
  name: string;
  pricePaisa: number;
  capacity: number;
  [key: string]: unknown;
};

export function listEvents() {
  return getStore().events as EventRecord[];
}

export function getEventBySlug(slug: string) {
  return getStore().events.find((event) => event.slug === slug || event.id === slug) as EventRecord | undefined;
}

export function createEvent(payload: Record<string, unknown>) {
  const event = {
    id: createId("evt"),
    organizationId: String(payload.organizationId ?? getStore().organizations[0]?.id ?? "org_demo"),
    slug: String(payload.slug ?? String(payload.title ?? "event").toLowerCase().replaceAll(" ", "-")),
    title: String(payload.title ?? "Untitled event"),
    description: String(payload.description ?? "New GatePass event"),
    eventType: String(payload.eventType ?? "public"),
    status: String(payload.status ?? "draft"),
    visibility: String(payload.visibility ?? "public"),
    venue: String(payload.venue ?? "Venue"),
    city: String(payload.city ?? "Mumbai"),
    startTime: String(payload.startTime ?? new Date().toISOString()),
    gpsRequired: Boolean(payload.gpsRequired ?? false),
  };
  getStore().events.push(event);
  void persistStoreRecord("events", event).catch((error) => console.error("Event persistence failed", error));
  return { event };
}

export function listTicketCategories(eventId?: string) {
  const categories = getStore().ticketCategories;
  return (eventId ? categories.filter((category) => category.eventId === eventId) : categories) as TicketCategoryRecord[];
}

export function createTicketCategory(payload: Record<string, unknown>) {
  const category = {
    id: createId("cat"),
    eventId: String(payload.eventId ?? getStore().events[0]?.id),
    name: String(payload.name ?? "General"),
    pricePaisa: Number(payload.pricePaisa ?? 0),
    capacity: Number(payload.capacity ?? 100),
  };
  getStore().ticketCategories.push(category);
  void persistStoreRecord("ticketCategories", category).catch((error) => console.error("Ticket category persistence failed", error));
  return { category };
}

export function publishEvent(id: string) {
  const event = getStore().events.find((entry) => entry.id === id) as EventRecord | undefined;
  if (event) event.status = "live";
  if (event) {
    void persistStoreUpdate("events", event).catch((error) => console.error("Event publish persistence failed", error));
  }
  return event;
}

export function updateEvent(id: string, payload: Record<string, unknown>) {
  const event = getStore().events.find((entry) => entry.id === id) as EventRecord | undefined;
  if (event) Object.assign(event, payload);
  if (event) {
    void persistStoreUpdate("events", event).catch((error) => console.error("Event update persistence failed", error));
  }
  return event;
}
